src/driver-sinowealth.c

```diff
@@ -0,0 +1,400 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+
+#include "libratbag-private.h"
+#include "libratbag-hidraw.h"
+
+#define SINOWEALTH_REPORT_ID_CONFIG 0x4
+#define SINOWEALTH_REPORT_ID_CMD 0x5
+#define SINOWEALTH_CMD_FIRMWARE_VERSION 0x1
+#define SINOWEALTH_CMD_GET_CONFIG 0x11
+#define SINOWEALTH_CONFIG_SIZE 520
+
+#define SINOWEALTH_XY_INDEPENDENT 0x80
+
+/* The PC software only goes down to 400, but PMW3360 doesn't care */
+#define SINOWEALTH_DPI_MIN 100
+#define SINOWEALTH_DPI_MAX 12000
+#define SINOWEALTH_DPI_STEP 100
+
+/* other models might have up to eight */
+#define SINOWEALTH_NUM_DPIS 6
+
+typedef struct __attribute__((packed)) {
+	uint8_t r, g, b;
+} RGB8;
+
+enum rgb_effect {
+	RGB_OFF = 0,
+	RGB_GLORIOUS = 0x1,   /* unicorn mode */
+	RGB_SINGLE = 0x2,     /* single constant color */
+	RGB_BREATHING = 0x5,  /* RGB breathing */
+	RGB_BREATHING7 = 0x3, /* breathing with seven colors */
+	RGB_BREATHING1 = 0xa, /* single color breathing */
+	RGB_TAIL = 0x4,       /* idk what this is supposed to be */
+	RGB_RAVE = 0x7,       /* ig */
+	RGB_WAVE = 0x9
+};
+
+struct sinowealth_config_report {
+	uint8_t report_id; /* SINOWEALTH_REPORT_ID_CONFIG */
+	uint8_t command_id;
+	uint8_t unk1;
+	uint8_t config_write;
+	/* always 0 when config is read from device,
+	 * has to be 0x7b when writing config to device
+	 */
+	uint8_t unk2[6];
+	uint8_t config;
+	/* 0x80 - SINOWEALTH_XY_INDEPENDENT */
+	uint8_t dpi_count:4;
+	uint8_t active_dpi:4;
+	uint8_t dpi_enabled;
+	/* bit set: disabled, unset: enabled
+	 * this structure has support for eight DPI slots,
+	 * but the glorious software only exposes six
+	 */
+	uint8_t dpi[16];
+	/* DPI/CPI is encoded in the way the PMW3360 sensor accepts it
+	 * value = (DPI - 100) / 100
+	 * If XY are identical, dpi[0-6] contain the sensitivities,
+	 * while in XY independent mode each entry takes two chars for X and Y.
+	 */
+	RGB8 dpi_color[8];
+
+	uint8_t rgb_effect;
+	/* see enum rgb_effect */
+
+	char glorious_mode;
+	/* 0x40 - brightness (constant)
+	 * 0x1/2/3 - speed
+	 */
+	char glorious_direction;
+
+	RGB8 single_color;
+
+	char breathing_mode;
+	/* 0x40 - brightness (constant)
+	 * 0x1/2/3 - speed
+	 */
+	char breathing_colorcount;
+	/* 7, constant */
+	RGB8 breathing_colors[7];
+
+	char tail_mode;
+	/* 0x10/20/30/40 - brightness
+	 * 0x1/2/3 - speed
+	 */
+
+	char rave_mode;
+	/* 0x10/20/30/40 - brightness
+	 * 0x1/2/3 - speed
+	 */
+	RGB8 rave_colors[2];
+
+	char wave_mode;
+	/* 0x10/20/30/40 - brightness
+	 * 0x1/2/3 - speed
+	 */
+
+	char breathing1_mode;
+	/* 0x1/2/3 - speed */
+	RGB8 breathing1_color;
+
+	char unk4;
+	char lift_off_distance;
+	/* 0x1 - 2 mm
+	 * 0x2 - 3 mm
+	 */
+} __attribute__((packed));
```

Comment:
Can you add a static assert for the size? This should make sure we don't compile with compilers which do not pack properly.

Example: https://github.com/libratbag/libratbag/blob/02575a90083b6cc3d5dd1a2d510acc34c4c24443/src/hidpp20.h#L904